<a href="https://colab.research.google.com/github/stzoozz/ece285-vision/blob/main/assignment4_release.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 4: Self-Attention for Vision

For this assignment, we're going to implement self-attention blocks in a convolutional neural network for CIFAR-10 Classification.

# Part I. Preparation

First, we load the CIFAR-10 dataset. This might take a couple minutes the first time you do it, but the files should stay cached after that.

In [ ]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader
from torch.utils.data import sampler

import torchvision.datasets as dset
import torchvision.transforms as T

import numpy as np

In [ ]:
NUM_TRAIN = 49000

# The torchvision.transforms package provides tools for preprocessing data
# and for performing data augmentation; here we set up a transform to
# preprocess the data by subtracting the mean RGB value and dividing by the
# standard deviation of each RGB value; we've hardcoded the mean and std.
transform = T.Compose([
                T.ToTensor(),
                T.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010))
            ])

# We set up a Dataset object for each split (train / val / test); Datasets load
# training examples one at a time, so we wrap each Dataset in a DataLoader which
# iterates through the Dataset and forms minibatches. We divide the CIFAR-10
# training set into train and val sets by passing a Sampler object to the
# DataLoader telling how it should sample from the underlying Dataset.
cifar10_train = dset.CIFAR10('./data/datasets', train=True, download=True,
                             transform=transform)
loader_train = DataLoader(cifar10_train, batch_size=64,
                          sampler=sampler.SubsetRandomSampler(range(NUM_TRAIN)))

cifar10_val = dset.CIFAR10('./data/datasets', train=True, download=True,
                           transform=transform)
loader_val = DataLoader(cifar10_val, batch_size=64,
                        sampler=sampler.SubsetRandomSampler(range(NUM_TRAIN, 50000)))

cifar10_test = dset.CIFAR10('./data/datasets', train=False, download=True,
                            transform=transform)
loader_test = DataLoader(cifar10_test, batch_size=64)

100%|██████████| 170M/170M [00:03<00:00, 43.1MB/s]


You have an option to **use GPU by setting the flag to True below**. It is not necessary to use GPU for this assignment. Note that if your computer does not have CUDA enabled, `torch.cuda.is_available()` will return False and this notebook will fallback to CPU mode.

The global variables `dtype` and `device` will control the data types throughout this assignment.

In [ ]:
USE_GPU = True

dtype = torch.float32 # we will be using float throughout this tutorial

if USE_GPU and torch.cuda.is_available():
    device = torch.device('cuda')
else:
    device = torch.device('cpu')

# Constant to control how frequently we print train loss
print_every = 100

print('using device:', device)

using device: cuda


## Flatten Function

In [ ]:
def flatten(x):
    N = x.shape[0] # read in N, C, H, W
    return x.view(N, -1)  # "flatten" the C * H * W values into a single vector per image

def test_flatten():
    x = torch.arange(12).view(2, 1, 3, 2)
    print('Before flattening: ', x)
    print('After flattening: ', flatten(x))

test_flatten()

Before flattening:  tensor([[[[ 0,  1],
          [ 2,  3],
          [ 4,  5]]],


        [[[ 6,  7],
          [ 8,  9],
          [10, 11]]]])
After flattening:  tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11]])


### Check Accuracy Function


In [ ]:
import torch.nn.functional as F  # useful stateless functions
def check_accuracy(loader, model):
    if loader.dataset.train:
        print('Checking accuracy on validation set')
    else:
        print('Checking accuracy on test set')
    num_correct = 0
    num_samples = 0
    model.eval()  # set model to evaluation mode
    with torch.no_grad():
        for x, y in loader:
            x = x.to(device=device, dtype=dtype)  # move to device, e.g. GPU
            y = y.to(device=device, dtype=torch.long)
            scores = model(x)
            _, preds = scores.max(1)
            num_correct += (preds == y).sum()
            num_samples += preds.size(0)
        acc = float(num_correct) / num_samples
        print('Got %d / %d correct (%.2f)' % (num_correct, num_samples, 100 * acc))
        return 100 * acc

### Training Loop

In [ ]:
def train(model, optimizer, epochs=1):
    """
    Train a model on CIFAR-10 using the PyTorch Module API.

    Inputs:
    - model: A PyTorch Module giving the model to train.
    - optimizer: An Optimizer object we will use to train the model
    - epochs: (Optional) A Python integer giving the number of epochs to train for

    Returns: Nothing, but prints model accuracies during training.
    """
    model = model.to(device=device)  # move the model parameters to CPU/GPU
    acc_max = 0
    for e in range(epochs):
        for t, (x, y) in enumerate(loader_train):

            model.train()  # put model to training mode
            x = x.to(device=device, dtype=dtype)  # move to device, e.g. GPU
            y = y.to(device=device, dtype=torch.long)

            scores = model(x)
            loss = F.cross_entropy(scores, y)

            # Zero out all of the gradients for the variables which the optimizer
            # will update.
            optimizer.zero_grad()

            # This is the backwards pass: compute the gradient of the loss with
            # respect to each  parameter of the model.
            loss.backward()

            # Actually update the parameters of the model using the gradients
            # computed by the backwards pass.
            optimizer.step()

            if t % print_every == 0:
                print('Epoch %d, Iteration %d, loss = %.4f' % (e, t, loss.item()))
                acc = check_accuracy(loader_val, model)
                if acc >= acc_max:
                    acc_max = acc
                print()
    print("Maximum accuracy attained: ", acc_max)

In [ ]:
# We need to wrap `flatten` function in a module in order to stack it
# in nn.Sequential
class Flatten(nn.Module):
    def forward(self, x):
        return flatten(x)

## Vanilla CNN; No Attention
We implement the vanilla architecture for you here. Do not modify the architecture. You will use the same architecture in the following parts. Do not modify the hyper-parameters.

In [ ]:
channel_1 = 64
channel_2 = 32
learning_rate = 1e-3
num_classes = 10

model = nn.Sequential(
    nn.Conv2d(3, channel_1, 3, padding=1, stride=1),
    nn.ReLU(),
    nn.Conv2d(channel_1, channel_2, 3, padding=1),
    nn.ReLU(),
    Flatten(),
    nn.Linear(channel_2*32*32, num_classes),
)

optimizer = optim.Adam(model.parameters(), lr=learning_rate)


train(model, optimizer, epochs=1)

Epoch 0, Iteration 0, loss = 2.3277
Checking accuracy on validation set
Got 138 / 1000 correct (13.80)

Epoch 0, Iteration 100, loss = 1.4879
Checking accuracy on validation set
Got 466 / 1000 correct (46.60)

Epoch 0, Iteration 200, loss = 1.3589
Checking accuracy on validation set
Got 509 / 1000 correct (50.90)

Epoch 0, Iteration 300, loss = 1.2044
Checking accuracy on validation set
Got 530 / 1000 correct (53.00)

Epoch 0, Iteration 400, loss = 1.2370
Checking accuracy on validation set
Got 553 / 1000 correct (55.30)

Epoch 0, Iteration 500, loss = 1.1319
Checking accuracy on validation set
Got 564 / 1000 correct (56.40)

Epoch 0, Iteration 600, loss = 1.1998
Checking accuracy on validation set
Got 585 / 1000 correct (58.50)

Epoch 0, Iteration 700, loss = 1.1249
Checking accuracy on validation set
Got 587 / 1000 correct (58.70)

Maximum accuracy attained:  58.699999999999996


## Test set -- run this only once

Now we test our model on the test set . Think about how this compares to your validation set accuracy.
You should be able to see atleast 55% accuracy

In [ ]:
vanillaModel = model
check_accuracy(loader_test, vanillaModel)


Checking accuracy on test set
Got 6008 / 10000 correct (60.08)


60.08

## Part II Self-Attention

In the next section, you will implement an Attention layer which you will then use within a convnet architecture defined above for cifar 10 classification task.

A self-attention layer is formulated as following:

Input: $X$ of shape $(H\times W, C)$

Query, key, value linear transforms are $W_Q$, $W_K$, $W_V$, of shape $(C, C)$. We implement these linear transforms as 1x1 convolutional layers of the same dimensions.

$XW_Q$, $XW_K$, $XW_V$, represent the output volumes when input X is passed through the transforms.


Self-Attention is given by the formula: $Attention(X) = X + Softmax(\frac{XW_Q(XW_K)^\top}{\sqrt{C}})XW_V$

### Inline Question 1: Self-Attention is equivalent to which of the following: (5 points)
1. K-means clustering <br />
2. Non-local means <br />
3. Residual Block <br />
4. Gaussian Blurring <br />

Your Answer: Non-local means

### Here you implement the Attention module, and run it in the next section (40 points)

In [ ]:
# Initialize the attention module as a nn.Module subclass
class Attention(nn.Module):
    def __init__(self, in_channels):
        super().__init__()

        # TODO: Implement the Key, Query and Value linear transforms as 1x1 convolutional layers
        # Hint: channel size remains constant throughout
        self.conv_query = nn.Conv2d(in_channels, in_channels, kernel_size=1)
        self.conv_key = nn.Conv2d(in_channels, in_channels, kernel_size=1)
        self.conv_value = nn.Conv2d(in_channels, in_channels, kernel_size=1)

    def forward(self, x):
        N, C, H, W = x.shape

        # TODO: Pass the input through conv_query, reshape the output volume to (N, C, H*W)
        q = self.conv_query(x)
        q = q.view(N, C, H * W).permute(0, 2, 1)
        # TODO: Pass the input through conv_key, reshape the output volume to (N, C, H*W)
        k = self.conv_key(x)
        k = k.view(N, C, H * W)
        # TODO: Pass the input through conv_value, reshape the output volume to (N, C, H*W)
        v = self.conv_value(x)
        v = v.view(N, C, H * W).permute(0, 2, 1)
        # TODO: Implement the above formula for attention using q, k, v, C
        # NOTE: The X in the formula is already added for you in the return line

        attention_scores = torch.bmm(q, k)
        attention_scores = attention_scores / (C ** 0.5)
        attention_weights = F.softmax(attention_scores, dim=-1)

        attention_output = torch.bmm(attention_weights, v)
        attention_output = attention_output.permute(0, 2, 1)

        # Reshape the output to (N, C, H, W) before adding to the input volume
        attention = attention_output.reshape(N, C, H, W)
        return x + attention

## Single Attention Block: Early attention; After the first conv layer. (10 points)

In [ ]:
channel_1 = 64
channel_2 = 32
learning_rate = 3e-4

# TODO: Use the above Attention module after the first Convolutional layer.
# Essentially the architecture should be [Conv->Relu->Attention->Relu->Conv->Relu->Linear]

model = nn.Sequential(
    nn.Conv2d(3, channel_1, 3, padding=1, stride=1),
    nn.ReLU(),
    Attention(in_channels=channel_1), # attention after first conv
    nn.ReLU(),
    nn.Conv2d(channel_1, channel_2, 3, padding=1),
    nn.ReLU(),
    Flatten(),
    nn.Linear(channel_2 * 32 * 32, 10)
)

optimizer = optim.Adam(model.parameters(), lr=learning_rate)

train(model, optimizer, epochs=10)

Epoch 0, Iteration 0, loss = 2.3105
Checking accuracy on validation set
Got 140 / 1000 correct (14.00)

Epoch 0, Iteration 100, loss = 1.6362
Checking accuracy on validation set
Got 435 / 1000 correct (43.50)

Epoch 0, Iteration 200, loss = 1.4686
Checking accuracy on validation set
Got 481 / 1000 correct (48.10)

Epoch 0, Iteration 300, loss = 1.2821
Checking accuracy on validation set
Got 540 / 1000 correct (54.00)

Epoch 0, Iteration 400, loss = 1.2726
Checking accuracy on validation set
Got 555 / 1000 correct (55.50)

Epoch 0, Iteration 500, loss = 1.2524
Checking accuracy on validation set
Got 554 / 1000 correct (55.40)

Epoch 0, Iteration 600, loss = 1.3331
Checking accuracy on validation set
Got 575 / 1000 correct (57.50)

Epoch 0, Iteration 700, loss = 1.2631
Checking accuracy on validation set
Got 585 / 1000 correct (58.50)

Epoch 1, Iteration 0, loss = 1.0244
Checking accuracy on validation set
Got 580 / 1000 correct (58.00)

Epoch 1, Iteration 100, loss = 0.9676
Checking acc

## Test set -- run this only once

Now we test our model on the test set . Think about how this compares to your validation set accuracy.
You should see improvement of about 2-3% over the vanilla convnet model. * Use this part to tune your Attention module and then move on to the next parts. *

In [ ]:
earlyAttention = model
check_accuracy(loader_test, earlyAttention)

Checking accuracy on test set
Got 6477 / 10000 correct (64.77)


64.77000000000001

## Single Attention Block: Late attention; After the second conv layer. (10 points)

In [ ]:
channel_1 = 64
channel_2 = 32
learning_rate = 3e-4

# TODO: Use the above Attention module after the Second Convolutional layer.
# Essentially the architecture should be [Conv->Relu->Conv->Relu->Attention->Relu->Linear]

model = nn.Sequential(
    nn.Conv2d(3, channel_1, 3, padding=1),
    nn.BatchNorm2d(channel_1),
    nn.ReLU(),

    nn.Conv2d(channel_1, channel_2, 3, padding=1),
    nn.BatchNorm2d(channel_2),
    nn.ReLU(),

    Attention(in_channels=channel_2),
    nn.ReLU(),

    Flatten(),
    nn.Linear(channel_2 * 32 * 32, 10)
)

optimizer = optim.Adam(model.parameters(), lr=learning_rate)

train(model, optimizer, epochs=10)

Epoch 0, Iteration 0, loss = 2.3764
Checking accuracy on validation set
Got 93 / 1000 correct (9.30)

Epoch 0, Iteration 100, loss = 1.8525
Checking accuracy on validation set
Got 453 / 1000 correct (45.30)

Epoch 0, Iteration 200, loss = 1.0844
Checking accuracy on validation set
Got 528 / 1000 correct (52.80)

Epoch 0, Iteration 300, loss = 1.5269
Checking accuracy on validation set
Got 535 / 1000 correct (53.50)

Epoch 0, Iteration 400, loss = 1.0713
Checking accuracy on validation set
Got 568 / 1000 correct (56.80)

Epoch 0, Iteration 500, loss = 1.1788
Checking accuracy on validation set
Got 579 / 1000 correct (57.90)

Epoch 0, Iteration 600, loss = 1.1484
Checking accuracy on validation set
Got 591 / 1000 correct (59.10)

Epoch 0, Iteration 700, loss = 0.9873
Checking accuracy on validation set
Got 623 / 1000 correct (62.30)

Epoch 1, Iteration 0, loss = 0.7901
Checking accuracy on validation set
Got 618 / 1000 correct (61.80)

Epoch 1, Iteration 100, loss = 0.7593
Checking accur

## Test set -- run this only once

Now we test our model on the test set . Think about how this compares to your validation set accuracy.

In [ ]:
lateAttention = model
check_accuracy(loader_test, lateAttention)

Checking accuracy on test set
Got 6352 / 10000 correct (63.52)


63.519999999999996

### Inline Question 2: Provide one example each of usage of self-attention and attention in computer vision. Explain the difference between the two. (5 points)


Your Answer:

- self-attention:

  vision transformers (ViT) - image patches attend to each other within the same image to capture long-range dependencies (e.g., dog head patch attending to tail patch).

- attention:

  image captioning - text decoder attends to relevant image regions when generating words (e.g., attending to frisbee region while outputting "frisbee").

- difference:

  - self-attention: within single input (intra-image)

  - attention: between different inputs (image→text cross-modal)
  
  - self-attention is special case where query/key/value come from same source.

## Double Attention Blocks: After conv layers 1 and 2 (10 points)

In [ ]:
channel_1 = 64
channel_2 = 32
learning_rate = 1e-3

# TODO: Use the above Attention module after the Second Convolutional layer.
# Essentially the architecture should be [Conv->Relu->Attention->Relu->Conv->Relu->Attention->Relu->Linear]

model = nn.Sequential(
    nn.Conv2d(3, channel_1, 3, padding=1),
    nn.BatchNorm2d(channel_1),
    nn.ReLU(),
    Attention(in_channels=channel_1),
    nn.ReLU(),

    nn.Conv2d(channel_1, channel_2, 3, padding=1),
    nn.BatchNorm2d(channel_2),
    nn.ReLU(),
    Attention(in_channels=channel_2),
    nn.ReLU(),

    Flatten(),
    nn.Dropout(0.3),
    nn.Linear(channel_2 * 32 * 32, 10)
)
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

train(model, optimizer, epochs=10)

Epoch 0, Iteration 0, loss = 2.3407
Checking accuracy on validation set
Got 119 / 1000 correct (11.90)

Epoch 0, Iteration 100, loss = 1.5102
Checking accuracy on validation set
Got 412 / 1000 correct (41.20)

Epoch 0, Iteration 200, loss = 1.4200
Checking accuracy on validation set
Got 516 / 1000 correct (51.60)

Epoch 0, Iteration 300, loss = 1.5608
Checking accuracy on validation set
Got 548 / 1000 correct (54.80)

Epoch 0, Iteration 400, loss = 1.0881
Checking accuracy on validation set
Got 544 / 1000 correct (54.40)

Epoch 0, Iteration 500, loss = 1.2188
Checking accuracy on validation set
Got 583 / 1000 correct (58.30)

Epoch 0, Iteration 600, loss = 1.1223
Checking accuracy on validation set
Got 618 / 1000 correct (61.80)

Epoch 0, Iteration 700, loss = 1.3264
Checking accuracy on validation set
Got 620 / 1000 correct (62.00)

Epoch 1, Iteration 0, loss = 1.0474
Checking accuracy on validation set
Got 614 / 1000 correct (61.40)

Epoch 1, Iteration 100, loss = 0.7106
Checking acc

## Test set -- run this only once

Now we test our model on the test set . Think about how this compares to your validation set accuracy.

In [ ]:
vanillaModel = model
check_accuracy(loader_test, vanillaModel)

Checking accuracy on test set
Got 6690 / 10000 correct (66.90)


66.9

## Resnet with Attention

Now we will experiment with applying attention within the Resnet10 architecture that we implemented in Homework 2. Please note that for a deeper model such as Resnet we do not expect significant improvements in performance with Attention

## Vanilla Resnet, No Attention

The architecture for Resnet is given below, please train it and evaluate it on the test set.

In [ ]:
import torch
import torch.nn as nn

class ResNet(nn.Module):

    def __init__(self, block, layers, img_channels=3, num_classes=100, batchnorm=False):
        super(ResNet, self).__init__() #layers = [1, 1, 1, 1]
        self.in_channels = 64
        self.conv1 = nn.Conv2d(img_channels, 64, kernel_size=7, stride=2, padding=3)
        self.bn1 = nn.BatchNorm2d(64)
        self.relu = nn.ReLU()
        self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
        self.batchnorm = batchnorm
        self.layer1 = self.make_layer(block, layers[0], out_channels=64, stride=1, batchnorm=batchnorm)
        self.layer2 = self.make_layer(block, layers[1], out_channels=128, stride=1, batchnorm=batchnorm)
        self.layer3 = self.make_layer(block, layers[2], out_channels=256, stride=1, batchnorm=batchnorm)
        self.layer4 = self.make_layer(block, layers[3], out_channels=512, stride=2, batchnorm=batchnorm)

        self.averagepool = nn.AdaptiveAvgPool2d((1, 1))
        self.fc = nn.Linear(512, num_classes)


    def forward(self, x):

        x = self.conv1(x)
        if self.batchnorm:
            x = self.bn1(x)
        x = self.relu(x)
        x = self.maxpool(x)
        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)
        x = self.averagepool(x)
        x = x.reshape(x.shape[0], -1)
        x = x.reshape(x.shape[0], -1)
        x = self.fc(x)

        return x




    def make_layer(self, block, num_blocks, out_channels, stride, batchnorm=False):
        downsampler = None
        layers = []
        if stride != 1 or self.in_channels != out_channels:
            downsampler = nn.Sequential(nn.Conv2d(self.in_channels, out_channels, kernel_size = 1, stride = stride), nn.BatchNorm2d(out_channels))

        layers.append(block(self.in_channels, out_channels, downsampler, stride, batchnorm=batchnorm))

        self.in_channels = out_channels

        for i in range(num_blocks - 1):
            layers.append(block(self.in_channels, out_channels))


        return nn.Sequential(*layers)

class block(nn.Module):

    def __init__(self, in_channels, out_channels, downsampler = None, stride = 1, batchnorm=False):

        super(block, self).__init__()
        self.conv1 = nn.Conv2d(in_channels, out_channels, kernel_size = 3, padding = 2)
        self.bn1 = nn.BatchNorm2d(out_channels)
        self.conv2 = nn.Conv2d(out_channels, out_channels, kernel_size = 3, stride = stride)
        self.bn2 = nn.BatchNorm2d(out_channels)
        self.downsampler = downsampler
        self.relu = nn.ReLU()
        self.batchnorm = batchnorm


    def forward(self, x):

        residual = x
        x = self.conv1(x)
        if self.batchnorm:
            x = self.bn1(x)
        x = self.relu(x)
        x = self.conv2(x)
        if self.batchnorm:
            x = self.bn2(x)
        x = self.relu(x)

        if self.downsampler:
            residual = self.downsampler(residual)

        return self.relu(residual + x)



def ResNet10(num_classes = 100, batchnorm= False):

    return ResNet(block, [1, 1, 1, 1], num_classes=num_classes, batchnorm=batchnorm)

## Test set -- run this only once

Now we test our model on the test set . Think about how this compares to your validation set accuracy.

In [ ]:
learning_rate = 1e-3

model = ResNet10()

optimizer = optim.Adam(model.parameters(), lr=learning_rate)

train(model, optimizer, epochs=10)

vanillaResnet = model
check_accuracy(loader_test, vanillaResnet)

Epoch 0, Iteration 0, loss = 4.7243
Checking accuracy on validation set
Got 106 / 1000 correct (10.60)

Epoch 0, Iteration 100, loss = 1.7463
Checking accuracy on validation set
Got 326 / 1000 correct (32.60)

Epoch 0, Iteration 200, loss = 1.4935
Checking accuracy on validation set
Got 418 / 1000 correct (41.80)

Epoch 0, Iteration 300, loss = 1.3473
Checking accuracy on validation set
Got 481 / 1000 correct (48.10)

Epoch 0, Iteration 400, loss = 1.1654
Checking accuracy on validation set
Got 451 / 1000 correct (45.10)

Epoch 0, Iteration 500, loss = 1.1752
Checking accuracy on validation set
Got 501 / 1000 correct (50.10)

Epoch 0, Iteration 600, loss = 1.2837
Checking accuracy on validation set
Got 575 / 1000 correct (57.50)

Epoch 0, Iteration 700, loss = 1.3234
Checking accuracy on validation set
Got 596 / 1000 correct (59.60)

Epoch 1, Iteration 0, loss = 0.9653
Checking accuracy on validation set
Got 599 / 1000 correct (59.90)

Epoch 1, Iteration 100, loss = 1.3107
Checking acc

74.89

## Resnet with Attention (5 points)

In [ ]:
## Resnet with Attention

learning_rate = 1e-3

# TODO: Use the above Attention module after the 2nd resnet block i.e. after self.layer2.
class ResNetWithAttention(nn.Module):
    def __init__(self, num_classes=10, batchnorm=True):
        super().__init__()
        self.resnet = ResNet(block, [1, 1, 1, 1], img_channels=3,
                             num_classes=num_classes, batchnorm=batchnorm)

        self.attention = Attention(in_channels=128)

    def forward(self, x):
        x = self.resnet.conv1(x)
        if self.resnet.batchnorm:
            x = self.resnet.bn1(x)
        x = self.resnet.relu(x)
        x = self.resnet.maxpool(x)

        x = self.resnet.layer1(x)
        x = self.resnet.layer2(x)

        x = self.attention(x)


        x = self.resnet.layer3(x)
        x = self.resnet.layer4(x)

        x = self.resnet.averagepool(x)
        x = x.reshape(x.shape[0], -1)
        x = self.resnet.fc(x)

        return x

# Create model and optimizer
model = ResNetWithAttention(num_classes=10, batchnorm=True)

optimizer = optim.Adam(model.parameters(), lr=learning_rate)

train(model, optimizer, epochs=10)

Epoch 0, Iteration 0, loss = 2.4149
Checking accuracy on validation set
Got 118 / 1000 correct (11.80)

Epoch 0, Iteration 100, loss = 1.5272
Checking accuracy on validation set
Got 367 / 1000 correct (36.70)

Epoch 0, Iteration 200, loss = 1.2662
Checking accuracy on validation set
Got 483 / 1000 correct (48.30)

Epoch 0, Iteration 300, loss = 1.2032
Checking accuracy on validation set
Got 534 / 1000 correct (53.40)

Epoch 0, Iteration 400, loss = 1.1105
Checking accuracy on validation set
Got 540 / 1000 correct (54.00)

Epoch 0, Iteration 500, loss = 1.1374
Checking accuracy on validation set
Got 585 / 1000 correct (58.50)

Epoch 0, Iteration 600, loss = 1.2507
Checking accuracy on validation set
Got 519 / 1000 correct (51.90)

Epoch 0, Iteration 700, loss = 1.1318
Checking accuracy on validation set
Got 579 / 1000 correct (57.90)

Epoch 1, Iteration 0, loss = 1.1879
Checking accuracy on validation set
Got 590 / 1000 correct (59.00)

Epoch 1, Iteration 100, loss = 0.8112
Checking acc

## Test set -- run this only once

Now we test our model on the test set . Think about how this compares to your validation set accuracy.

In [ ]:
AttentionResnet = model
check_accuracy(loader_test, AttentionResnet)

Checking accuracy on test set
Got 7744 / 10000 correct (77.44)


77.44

## Inline Question 3: Rank the above models based on their performance on test dataset (15 points)
( You are encouraged to run each of the experiments (training) at
least 3 times to get an average estimate )

Report the test accuracies alongside the model names.

1. Vanilla CNN, No Attention (60.08%)
2. Vanilla CNN Single Attention Block: Late attention; After the second conv layer (63.52%)
3. Vanilla CNN Single Attention Block: Early attention; After the first conv layer (64.78%)
4. Vanilla CNN Double Attention Blocks: After conv layers 1 and 2 (66.9%)
5. Vanilla Resnet, No Attention (74.89%)
6. Vanilla Resnet with Attention (77.44%)

### Bonus Question (Ungraded): Can you give a possible explanation that supports the rankings?
Your Answer: better architecture (ResNet) + well-placed attention = best performance.